File: 198-house-robber/house-robber.js
/**
 * @param {number[]} nums
 * @return {number}
 */
const solve = (nums, n, memo = {}) => {
    if (n in memo) return memo[n];
    if (n < 0) return 0;
    if (n === 0) return nums[0];

    let loots = nums[n] + solve(nums, n - 2, memo);
    let not_loots = solve(nums, n - 1, memo);

    memo[n] = Math.max(loots, not_loots);
    return memo[n];
}

var rob = function(nums) {
    return solve(nums, nums.length-1);
};

File: 198-house-robber/house-robber.cpp
class Solution {
public:
    int solve(vector<int> &nums, int index, vector<int> &memo){
        if (index == 0) return nums[0];
        if (index < 0) return 0;
        if (memo[index] != -1) return memo[index];
        int robs = nums[index] + solve(nums, index - 2, memo);
        int not_robs = solve(nums, index - 1, memo);

        memo[index] = max(robs, not_robs);
        return memo[index];
    }
    int rob(vector<int>& nums) {
        vector<int> memo(nums.size(), -1);
        return solve(nums, nums.size()-1, memo);
    }
};